File: java/src/test/java/com/adaptionsoft/games/trivia/ApprovalGame.java
package com.adaptionsoft.games.trivia;

import com.adaptionsoft.games.uglytrivia.Game;
import org.approvaltests.Approvals;

public class ApprovalGame {
    private final BufferPresenter bufferPresenter;
    private final Dice tenDice;
    private final Dice sixDice;

    private ApprovalGame(long randomSeed) {
        tenDice = new Dice(4, randomSeed);
        sixDice = new Dice(10, randomSeed);
        bufferPresenter = new BufferPresenter();
    }

    public static ApprovalGame create(long randomSeed) {
        return new ApprovalGame(randomSeed);
    }

    public void runTestScenario() {
        final Game aGame = buildGame();

        runGame(aGame);

        approveMessages();
    }

    private void approveMessages() {
        Approvals.verifyAll("", bufferPresenter.getBufferedMessages().toArray());
    }


    private void runGame(Game aGame) {
        do {
            aGame.roll(sixDice.roll());
        } while (determineAnswer(aGame));
    }

    private boolean determineAnswer(Game aGame) {
        if (tenDice.roll() == 7) {
            return aGame.wrongAnswer();
        } else {
            return aGame.wasCorrectlyAnswered();
        }
    }


    private Game buildGame() {
        final Game aGame = new Game(bufferPresenter);
        aGame.add("Jan");
        aGame.add("Piet");
        aGame.add("Joris");
        aGame.add("Korneel");
        return aGame;
    }
}

File: kotlin/src/test/kotlin/com/adaptionsoft/games/trivia/ApprovalGame.kt
package com.adaptionsoft.games.trivia


import com.adaptionsoft.games.uglytrivia.*
import org.approvaltests.*

internal class ApprovalGame private constructor(randomSeed: Long) {
    private val bufferPresenter: BufferPresenter
    private val fourDice: Dice
    private val sixDice: Dice

    init {
        fourDice = Dice(4, randomSeed)
        sixDice = Dice(6, randomSeed)
        bufferPresenter = BufferPresenter()
    }

    fun runTestScenario() {
        val aGame = buildGame()
        runGame(aGame)
        approveMessages()
    }

    private fun approveMessages() {
        Approvals.verifyAll<Any>("", bufferPresenter.bufferedMessages.toTypedArray())
    }

    private fun runGame(aGame: Game) {
        do {
            aGame.roll(sixDice.roll())
        } while (determineAnswer(aGame))
    }

    private fun determineAnswer(aGame: Game): Boolean {
        return if (fourDice.roll() == 4) {
            aGame.wrongAnswer()
        } else {
            aGame.wasCorrectlyAnswered()
        }
    }

    private fun buildGame(): Game {
        val aGame = Game(bufferPresenter)
        aGame.add("Jan")
        aGame.add("Piet")
        aGame.add("Joris")
        aGame.add("Korneel")
        return aGame
    }

    companion object {
        @JvmStatic
        fun create(randomSeed: Long): ApprovalGame {
            return ApprovalGame(randomSeed)
        }
    }
}